gsettings: add page (#6251)
tldr-pages/tldr: pages/linux/gsettings.md

# gsettings

> > query and modify dconf settings with schema validation

Reviewer: What in the world is a schema validation, other than https://duckduckgo.com/?t=ffab&q=schema+validation 

Is that the average user will understand, because I don't think so.

Author: `gsettings set` will fail if you try to write a key that shouldn't exist, or supply a value that isn't in the range allowed by the schema. `dconf write` doesn't know about schemas and will let you write anything.

Author: I extended the `gsettings set` a little to mention this.

Reviewer: I don't think its required or could just be shortened to `valid input` 

cc @SethFalco @navarroaxel

Reviewer 2: I think most developers, sysadmins, and power users will probably understand or infer it.
It'd probably be fair to assume others might have to look it up.

I have no opinion on if it should be changed, though. Subjectively, I like it. ^-^'

Both are perfectly valid options, they're just different levels of explicitness:
`valid input` - Input is being validated _somehow_.
`schema validation` - Input is being validated against a schema. (A data model definition, like [JSON Schema](https://json-schema.org/) or [XSD](https://en.wikipedia.org/wiki/XML_Schema_(W3C)).)

Reviewer 3: (https://github.com/tldr-pages/tldr/pull/6251#discussion_r673337616)

> > Query and modify dconf settings with schema validation.

Reviewer 3: I'm not sure if we even need to be that specific. Reading that it's not obviously clear what it does (in my opinion). How about just:
> GSettings configuration tool

> > More information: <https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/using_the_desktop_environment_in_rhel_8/configuring-gnome-at-low-level_using-the-desktop-environment-in-rhel-8#using-gsettings-command_configuring-gnome-at-low-level>.

Reviewer: This link is way to long I feel.

And from what I see, maybe we should use the manned.org link

Author: It is unfortunately very long, but the Redhat documentation does contain examples and links to learn more about the workings of dconf and gsettings.  Also, I feel that anyone who can `tldr gsettings` can just as well `man gsettings` on their own machine.  And may in fact have just *come* from the manpage.

Reviewer: I'll look for an official site rather than a guide, if you can find one though since you know more about the topic, that would be better

Reviewer 3: It's unfortunate that it's so long, but it seems quite good and far more usable than the official GNOME documentation or the man page.

Author: One option might be the developer documentation for GSettings itself, which is part of GLib.  The url is not overlong: https://developer.gnome.org/gio/stable/settings.html.  It doesn't really talk about the command line tool `gsettings` at all, but AFAICT, `gsettings` is a fairly thin wrapper around a few of the functions there, and if you do a deep-dive on it you'll probably get a better understanding than the man page.

Reviewer 3: I don't think that page would be helpful for anyone wanting to use the command, but the `redhat` page still links to the official docs.

> - Dump all (non-relocatable) schemas, keys, and values:

Reviewer: By `dump` what do you mean, does it give complete raw data or is it more 'sincere' in how it gives the info

We normally use Display or Print

Author: I have made this change, using "display" where the output is more useful for human consumption, and "print" where the output is directly machine-readable and probably less likely to change and break one's scripts in the future.

Author: It is indeed not entirely complete (doesn't contain the output of `range` or `describe`, for example), and not importable by anything AFAIK. But it is useful for piping to grep if you forget the exact name of a schema or key.

> - Display all (non-relocatable) schemas, keys, and values:

Reviewer 3: Is there maybe an easier term or a way to describe `non-relocatable`?

Author: What I was trying to draw attention to is that `list-recursively` will not actually show *everything*.

What [the documentation](https://developer.gnome.org/gio/stable/GSettings.html) has to say about relocatable schemas is:

>A relocatable schema is one with no path attribute specified on its <schema> element.

[...]

>For example, a relocatable schema could be used to store geometry information for different windows in an application. If the schema ID was org.foo.MyApp.Window, it could be instantiated for paths /org/foo/MyApp/main/, /org/foo/MyApp/document-1/, /org/foo/MyApp/document-2/, etc.

[And elsewhere:](https://developer.gnome.org/gio/stable/gio-GSettingsSchema-GSettingsSchemaSource.html#g-settings-schema-get-path)

>Schemas may be single-instance or relocatable. Single-instance schemas correspond to exactly one set of keys in the backend database: those located at the path returned by this function.

>Relocatable schemas can be referenced by other schemas and can therefore describe multiple sets of keys at different locations. For relocatable schemas, this function will return `NULL`.

The main distinction seems to be that in order to operate on relocatable schemas, you need to know the GSettings schema id AND the dconf path.  `gsettings list-relocatable-schemas` can tell you the ids, but IDK how to get the paths where those schemas might be used.  [According to the discussion here](https://github.com/mate-desktop/mate-panel/issues/675), the `dconf-editor` GUI tool had a bear of a time trying to map the other way (from path to schema id).

It seems like most uses of the `gsettings` tool are, "paste this in your terminal to change such and such setting not exposed in the GUI," or new computer setup automation, which largely don't involve relocatable schemas.

Author: In most recent version, I have re-ordered the examples to put the subcommands with scripting uses, `set`/`get`/`reset`, come first and the "exploratory" ones, including `list-recursively`, are farther down.  This makes the stuff about non-relocatable vs relocatable schemas less prominent, so hopefully users are less likely to get the impression they need to understand what that means to use the tool.

But I do think it should be there somewhere, as a warning that `list-recursively` does not display everything.

`gsettings list-recursively`

> - Display all keys and values (default if not set) from one schema:

Reviewer 3: What does the `default` mean here? Is this behaviour default or does it use a default schema?

Author: The default value for that key.

AFAICT, GSettings uses a dconf database backend that only stores changes from the default values specified in the schema definition.  Schema definitions are found in `/usr/share/glib-2.0/schemas`.

So:

```
$ gsettings reset org.gnome.settings-daemon.plugins.xsettings hinting
$ gsettings get org.gnome.settings-daemon.plugins.xsettings hinting
'slight'
$ dconf read /org/gnome/settings-daemon/plugins/xsettings/hinting
$ # (no output)
```

But if we explicitly set that key, even to one that matches the default...

```
$ gsettings set org.gnome.settings-daemon.plugins.xsettings hinting slight
$ gsettings get org.gnome.settings-daemon.plugins.xsettings hinting
'slight'
$ dconf read /org/gnome/settings-daemon/plugins/xsettings/hinting
'slight'
```

It is stored in the dconf database.

(And incidentally, we have triggered [a very old cairo bug](https://gitlab.freedesktop.org/cairo/cairo/-/issues/23).)

> - Print the value of a key, or the schema-provided default if the key has not been set in dconf:

Reviewer 3: ```suggestion
- Print the value of a key or the schema-provided default if the key has not been set in `dconf`:
```

`gsettings list-recursively {{org.example.schema}}`

> - Set the value of a key. Fails if key doesn't exist or value is outside range:

Reviewer 3: ```suggestion
- Set the value of a key. Fails if the key doesn't exist or the value is out of range:
```

> `gsettings set {{org.example.schema}} {{example-key}} {{value}}`

Reviewer 3: Asking fellow maintainers: Should we use `example_key` here like we do everywhere or should we use the formatting style of a gsettings key (e.g. `always-show-log-out`)?

Author: For what it's worth, [the docs](https://developer.gnome.org/gio/stable/GSettings.html) do specifically say:

>Key names are restricted to lowercase characters, numbers and '-'.

Reviewer 3: That's why I'm not sure. Independent of what the command uses, we generally use snake_case for arguments.

> - Un-set a key, so that its schema default value will be used:

Reviewer 2: Sorry to be a pain!

```suggestion
- Unset a key, so that its schema default value will be used:
```

* https://www.collinsdictionary.com/dictionary/english/unset
* https://www.lexico.com/definition/unset
* https://www.merriam-webster.com/dictionary/unset

"Unset" is one word! Other than that I think this PR looks great!

- Un-set a key, so that its schema default value will be used:

`gsettings reset {{org.example.schema}} {{example-key}}`

> - Print the value of a key, or schema default if not set (unlike `dconf read`):

Reviewer 3: Not sure about this comma here

`gsetings get {{org.example.schema}} {{example-key}}`

> - Display schema-allowed values for a key (helpful with enum keys, less so for others):

Reviewer 3: ```suggestion
- Display schema-allowed values for a key (helpful with enum keys):
```
Not sure this is needed. Saying that something's useful for one thing implies that it's not as useful for others.

`gsettings range {{org.example.schema}} {{example-key}}`

> - Display the human-readable description for a key:

Reviewer 3: ```suggestion
- Display the human-readable description of a key:
```
